frame/ethereum/src/lib.rs

```diff
@@ -145,22 +147,9 @@ decl_module! {
 		fn transact(origin, transaction: ethereum::Transaction) {
 			ensure_none(origin)?;
 
-			ensure!(
-				transaction.signature.chain_id().unwrap_or_default() == T::ChainId::get(),
-				Error::<T>::InvalidChainId
-			);
-			let mut sig = [0u8; 65];
-			let mut msg = [0u8; 32];
-			sig[0..32].copy_from_slice(&transaction.signature.r()[..]);
-			sig[32..64].copy_from_slice(&transaction.signature.s()[..]);
-			sig[64] = transaction.signature.standard_v();
-			msg.copy_from_slice(&transaction.message_hash(Some(T::ChainId::get()))[..]);
-
-			let pubkey = sp_io::crypto::secp256k1_ecdsa_recover(&sig, &msg)
-				.map_err(|_| "Recover public key failed")?;
-			let source = H160::from(H256::from_slice(Keccak256::digest(&pubkey).as_slice()));
-
-			Self::execute(source, transaction);
+			let source = Self::recover_signer(&transaction).ok_or_else(|| Error::<T>::InvalidSignature)?;
```

Comment:
not sure if we can avoid this extra `recover_signer`

Reply:
Can't think of a way to do it either.